src/binding/c/attr_api.txt

```diff
@@ -0,0 +1,202 @@
+    .polymorph: MPIR_Attr_type attr_type=MPIR_ATTR_PTR
+    .extra: ignore_revoke
+
+MPI_Comm_set_attr:
+    .desc: Stores attribute value associated with a key
+    .seealso: MPI_Comm_create_keyval, MPI_Comm_delete_attr
+    .polymorph: MPIR_Attr_type attr_type=MPIR_ATTR_PTR
+    .extra: ignore_revoke
+/*
+    Notes:
+    Values of the permanent attributes 'MPI_TAG_UB', 'MPI_HOST', 'MPI_IO',
+    'MPI_WTIME_IS_GLOBAL', 'MPI_UNIVERSE_SIZE', 'MPI_LASTUSEDCODE', and
+    'MPI_APPNUM' may not be changed.
+
+    The type of the attribute value depends on whether C, C++, or Fortran
+    is being used.
+    In C and C++, an attribute value is a pointer ('void *'); in Fortran, it is an
+    address-sized integer.
+
+    If an attribute is already present, the delete function (specified when the
+    corresponding keyval was created) will be called.
+*/
+
+MPI_Keyval_create:
+    .desc: Greates a new attribute key
+    .replace: deprecated with MPI_Comm_create_keyval
+    .seealso: MPI_Keyval_free, MPI_Comm_create_keyval
+/*
+    Notes:
+    Key values are global (available for any and all communicators).
+
+    There are subtle differences between C and Fortran that require that the
+    copy_fn be written in the same language that 'MPI_Keyval_create'
+    is called from.
+    This should not be a problem for most users; only programmers using both
+    Fortran and C in the same program need to be sure that they follow this rule.
+*/
+
+MPI_Keyval_free:
+    .desc: Frees an attribute key for communicators
+    .seealso: MPI_Keyval_create, MPI_Comm_free_keyval
+    .replace: deprecated with MPI_Comm_free_keyval
+
+MPI_Type_create_keyval:
+    .desc: Create an attribute keyval for MPI datatypes
+/*
+    Notes:
+
+       Default copy and delete functions are available.  These are
+    + MPI_TYPE_NULL_COPY_FN   - empty copy function
+    . MPI_TYPE_NULL_DELETE_FN - empty delete function
+    - MPI_TYPE_DUP_FN         - simple dup function
+
+    .N AttrErrReturn
+*/
+
+MPI_Type_delete_attr:
+    .desc: Deletes an attribute value associated with a key on
+
+MPI_Type_free_keyval:
+    .desc: Frees an attribute key for datatypes
+{
+    MPIR_free_keyval(type_keyval_ptr);
+    *type_keyval = MPI_KEYVAL_INVALID;
+}
+
+MPI_Type_get_attr:
+    .desc: Retrieves attribute value by key
+    .polymorph: MPIR_Attr_type attr_type=MPIR_ATTR_PTR
+
+MPI_Type_set_attr:
+    .desc: Stores attribute value associated with a key
+    .skip: ThreadSafe
+    .polymorph: MPIR_Attr_type attr_type=MPIR_ATTR_PTR
+/*
+    Notes:
+
+    The type of the attribute value depends on whether C or Fortran is being used.
+    In C, an attribute value is a pointer ('void *'); in Fortran, it is an
+    address-sized integer.
+
+    If an attribute is already present, the delete function (specified when the
+    corresponding keyval was created) will be called.
+*/
+
+MPI_Win_create_keyval:
+    .desc: Create an attribute keyval for MPI window objects
```

Comment:
Notes:
```
   Notes:	
   Default copy and delete functions are available.  These are	
+ MPI_WIN_NULL_COPY_FN   - empty copy function	
. MPI_WIN_NULL_DELETE_FN - empty delete function	
- MPI_WIN_DUP_FN         - simple dup function
```

Maybe it is autogenerated, but then `MPI_Type_create_keyval()` has a redundant note.

Reply:
I believe that inconsistency was there in the current code. I agree with you that we should make them consistent and probably this is the time.

Reply:
fixed in #4955